### PR TITLE
Nadav/handle route exports

### PR DESCRIPTION
**Pull request by nadav-ab**

### CODE REVIEW

File: packages/define-remix-app/test/test-cases/roots.ts [hunk 1]
```diff
@@ -375,3 +446,76 @@ export const actionPage = (name: string) =>
...
+    export const action = async ({ request, params }: LoaderFunctionArgs) => {
+        const nickname = params.nickname;
+        const existing = userNames.has(nickname);
+        const body = await request.formData();
+        const newEmail = body.get('email');
+        const newFullName = body.get('fullName');
+        if (newEmail && newFullName) {
+            userNames.set(nickname, { email: newEmail, fullName: newFullName });
+            return json({
+                message: existing ? 'User updated' : 'User created',
```
Comment (idoros) [inline packages/define-remix-app/test/test-cases/roots.ts:471]: No test for "User updated" or "User does not exist"? maybe simplify the fixture?

File: packages/define-remix-app/src/manifest-to-router.tsx [hunk 2]
```diff
@@ -248,10 +238,35 @@ function lazyCompAndLoader(
...
     prevUri: { current: string },
     callServerMethod: (filePath: string, methodName: string, args: unknown[]) => Promise<unknown>,
-) {
+): RouteObject {
+    const { handle, setHandle } = createHandleProxy();
+    const { linksWrapper, setLinks } = createLinksProxy();
+
+    const importModuleAndUpdate: DynamicImport = (filePath, cb) => {
+        const { moduleResults, dispose } = requireModule(filePath, (newResults) => {
+            setHandle((newResults.results as { handle?: unknown }).handle);
```
Comment (idoros) [inline packages/define-remix-app/src/manifest-to-router.tsx:247]: Should be ok as the module system returns `{}` on error, but maybe the `DynamicImport` should be generic and the remix integration could expect it's own module interface.

File: packages/define-remix-app/src/manifest-to-router.tsx [hunk 3]
```diff
@@ -248,10 +238,35 @@ function lazyCompAndLoader(
...
+            const linksFunction = (newResults.results as { links?: LinksFunction }).links;
+            if(linksFunction){
+                setLinks(linksFunction);
```
Comment (idoros) [inline packages/define-remix-app/src/manifest-to-router.tsx:250]: Wouldn't the condition here prevent links from being removed?

File: packages/define-remix-app/src/manifest-to-router.tsx [hunk 4]
```diff
@@ -275,15 +290,40 @@ function lazyCompAndLoader(
...
+    };
+    const loader: LoaderFunction | undefined = exportNames.includes('clientLoader')
         ? async ({ params, request }) => {
-              const res = await callServerMethod(filePath, 'loader', [
-                  { params, request: await serializeRequest(request) },
-              ]);
-              return isSerializedResponse(res) ? deserializeResponse(res) : res;
+              const lastResults: { current: IResults<unknown> } = { current: { status: 'loading', results: null } };
+              const { moduleResults } = importModuleAndUpdate(filePath, (updated) => {
+                  lastResults.current = updated;
```
Comment (idoros) [inline packages/define-remix-app/src/manifest-to-router.tsx:305]: What handles an error in loader after the loader updates into an error state?

**Review by idoros — COMMENTED** [review]
This isn't new to the code, but this PR introduces many additional listeners to the module system. I'm concerned that these listeners are not being properly disposed of, which could lead to accumulating observers on the module system and potential memory leaks.

File: packages/define-remix-app/test/test-cases/roots.ts [hunk 5]
```diff
@@ -229,7 +247,60 @@ export const loaderPage = (name: string, message: string) =>
...
+    }
+`);
 
+export const clientLoaderWithFallbackPage = (name: string,  clientMessage: string) =>
+    transformTsx(`
+    import React from 'react';
+    import { Outlet, useLoaderData } from '@remix-run/react';
+    const loaderPromise = new Promise((resolve) => {
+        const globalListener = () => {
+            debugger;
```
[the remark this answers is not in the document]
Comment (nadav-ab) [reply packages/define-remix-app/test/test-cases/roots.ts:285]: done

File: packages/define-remix-app/src/manifest-to-router.tsx (hunk 4 above)
Comment (nadav-ab) [reply packages/define-remix-app/src/manifest-to-router.tsx:305]: i dont understand the question

File: packages/define-remix-app/src/manifest-to-router.tsx (hunk 3 above)
Comment (nadav-ab) [reply packages/define-remix-app/src/manifest-to-router.tsx:250]: actually, the manifest changes when the exportnames change, revoking the memo, adding a test

File: packages/define-remix-app/src/manifest-to-router.tsx (hunk 4 above)
Comment (idoros) [reply packages/define-remix-app/src/manifest-to-router.tsx:305]: If we start with a valid page that has a loader and then it becomes invalid, then the `lastResults` reference the stale previous value. Maybe it doesn't matter.

File: packages/define-remix-app/src/manifest-to-router.tsx (hunk 2 above)
Comment (nadav-ab) [reply packages/define-remix-app/src/manifest-to-router.tsx:247]: maybe but not in this PR

File: packages/define-remix-app/src/manifest-to-router.tsx (hunk 3 above)
Comment (nadav-ab) [reply packages/define-remix-app/src/manifest-to-router.tsx:250]: added a test and fix

**Pull request merged**